### PR TITLE
Speed Increase

### PR DESCRIPTION
**Pull request by millardjn**
Hello, I have made some changes to increase the performance of the calculations:
 - Refer to HyperEdges by their position/index rather than passing around SmallVecs
 - Move ".contains()" calls to the setup, and cache results
 - Minimise recalculation of "equation_lhs()" in the gradient as much as possible.
 - Calculate the cluster probabilities in two halves, only one of which needs to be recalculated at a time.

This brought the surface_code calculation down to 2 seconds on my computer.

### CODE REVIEW

**Comment by millardjn** [comment]
It might be possible to improve the solution accuracy by using logits as the solver parameters to smoothly clamp the probabilities to between 0 and 1. I haven't tried it here, but it is a common way to avoid negative values in the solution if they are not wanted.

**Comment by inmzhang** [comment]
Thank you for providing so much guidance and assistance; it has truly helped me a lot.

**Comment by millardjn** [comment]
Hi, No worries! Sorry for responding so slow, work got very busy. There were a couple of points I wanted to handover before I forget:

- Nonlinear conjugate gradient requires a line search that nearly runs to completion (c2 = 0.05 rather than 0.9) to be efficient. This has to do with how conjugate gradient method differs from regular gradient decent. I had forgotten this when I first looked at 
- I was wrong when I initially said that I didn't expect poor conditioning. Given how much LBFGS and BFGS outperform NCG it is clear that having access to a hessian or an approximate hessian is very beneficial for this type of problem.

You clearly work on interesting things, so feel free to let me know if you'd like a hand with similar work in rust. All the best.
